### PR TITLE
feat: resave to svg if you opened an svg with a scene in it

### PR DESCRIPTION
**Pull request by alsuren**
This is a sketch of how I might go about fixing https://github.com/excalidraw/excalidraw/issues/2261 

I ended up doing a bunch of refactors while I was exploring the codebase. Some of them are valuable, but not all of them are needed.

It would be nice to have some feedback on whether this is a reasonable approach, but I will probably not try to get this merged. It will be better to close this PR, and split out a series of smaller PRs with refactorings, followed by a PR that implements the feature cleanly.

Summary of changes, in the order they appear as you see them in the PR diff view:

src/actions/actionExport.tsx
- [keep - last PR] rename loadFromJSON, saveAsJSON to loadFromFilesystem, saveToFilesystem
  - [ ] Maybe saveSceneToFilesystem, to be more explicit?
- [scrap]: pointless refactor in actionSaveScene

src/appState.ts:
- [keep - last PR] Add saveType to appState (svg | png | excalidraw)

src/clipboard.ts (also src/data/json.ts):
- [keep - first PR] Try to pass `string` or `Blob` around in the export modules, rather than Canvas or HtmlSvgElement or whatever

src/components/ExportDialog.tsx:
- [hack] export defaultScale  from ExportDialog for use in saveToFilesystem
  - [ ] (should really save this into the svg file and load it if possible?)

src/components/LayerUI.tsx:
- [keep - first PR] add exportEmbedScene as an option to the serialisation functions, rather than plucking it out of appState
  - [ ] need to finish this refactor in a few places

src/data/blob.ts:
- [keep - last PR] infer a saveType (svg | png | excalidraw) when loading scene from filesystem and add it to appState

src/data/image.ts:
- [keep] explicit return type for encodePngMetadata() (should we have a linter rule about this?)

src/data/index.ts:
- [scrap] split out `exportToSVGForReal()` and `exportToPNGForReal()` - This is what happens when you try to refactor bottom-up rather than top-down. When I tried again top-down, from saveSceneToFilesystem(), I realised that it was much nicer to clean up the serlialisation helpers.
- [keep - first PR] serializeToSvg() helper that returns a string
  - [ ] rename to serializeToSvgString() or somehow work out whether it's reasonable to return a Blob, for symmetry with below?
- [keep - first PR] serializeAsPngBlob helper that wraps exportToCanvas() and takes care of deleting the 
- [scrap] messing about with how the export functions call fileSave - this was definitely a bad idea
- [???] remove really strange thing where we compare tempCanvas with [app.]canvas. What is actually going on here?

src/data/json.ts:
- [ ] rename this file to something else, now that it doesn't just deal with the .excalidraw json format?
- [keep - last PR] switch on saveType to decide how to serialise the scene
  - [ ] how to deal with NonDeletedExcalidrawElement vs ExcalidrawElement?
  - [ ] audit the serializeTo*() parameters, and work out how we should deal with each of these
  - [ ] explicitly think about how SaveAs should work now

Misc:
- [ ] [first PR] I've made a bunch of changes to where we are importing/using various functions. It would be good to look over the dependencies between modules, and decide whether we want to shuffle things about.
- [ ] write some tests, maybe?

The call tree for exportToCanvas is here (with scene embedded):

status-quo

I would like it to look like this:

plan

I think that ??? will probably be called resaveAsImageWithScene() or something.

### CODE REVIEW

File: src/scene/export.ts [hunk 1]
```diff
@@ -86,6 +88,39 @@ export const exportToCanvas = (
   return tempCanvas;
 };
 
+export const serializeAsPngBlob = async (
+  elements: readonly NonDeletedExcalidrawElement[],
+  appState: AppState,
+  // FIXME: extract options type definition or inline exportToCanvas()?
+  options: {
```
Comment (ceddlyburge) [inline src/scene/export.ts:95]: This is already donw I think, and can become ExportOptions?

File: src/scene/export.ts [hunk 2]
```diff
@@ -228,3 +263,17 @@ export const getExportSize = (
 
   return [width, height];
 };
+
+export const serializeToSvg = (
+  elements: readonly NonDeletedExcalidrawElement[],
+  // FIXME: extract a type for this, or delete exportToSvg and inline?
```
Comment (ceddlyburge) [inline src/scene/export.ts:269]: This type does seem to appear in multiple placfes, so probably worth explicitly declaring it.

File: src/data/index.ts [hunk 3]
```diff
@@ -1,20 +1,47 @@
...
+    await exportToSVGForReal(type, elements, appState, canvas, options);
+  } else if (type === "png" || type === "clipboard") {
+    exportToPNGForReal(type, elements, appState, canvas, options);
+  }
+};
+
+const exportToSVGForReal = async (
   type: ExportType,
   elements: readonly NonDeletedExcalidrawElement[],
   appState: AppState,
```
[the remark this answers is not in the document]
Comment (alsuren) [reply src/data/index.ts:47]: yeah, that's probably a good idea. I'm still not sure what the philosophy is with AppState, and whether making things nested would be sensible. At the moment, it seems to be used for tracking the state of tickyboxes and such in the UI (in the react controlled-component style), and then it gets flung around all over the place and everything reaches in and plucks individual keys out of it.

File: src/data/index.ts [hunk 4]
```diff
@@ -26,86 +53,97 @@ export const exportCanvas = async (
...
   }
+};
 
-  const tempCanvas = exportToCanvas(elements, appState, {
+const exportToPNGForReal = async (
+  type: ExportType,
+  elements: readonly NonDeletedExcalidrawElement[],
+  appState: AppState,
+  canvas: HTMLCanvasElement,
+  {
```
[the remark this answers is not in the document]
Comment (alsuren) [reply src/data/index.ts:98]: Yeah, this definitely needs cleaning up. In practice I might even flatten this function back into exportCanvas(), now that I've cleaved off the `serializeAsPngBlob()` helper, and realised that I don't need to use this helper anywhere else.

File: src/data/json.ts [hunk 5]
```diff
@@ -24,15 +29,78 @@ export const serializeAsJSON = (
...
+    }
+    case "png": {
+      // FIXME: I guess we should do a garbage collect in the layer above,
+      // and pass NonDeletedExcalidrawElement[] into this function?
+      // For now, I'm just going to hack it.
+      return await serializeAsPngBlob(
+        elements as NonDeletedExcalidrawElement[],
+        appState,
+        {
+          // FIXME: I'm not sure if we can expect these pieces of appState
```
[the remark this answers is not in the document]
Comment (alsuren) [reply src/data/json.ts:70]: serializeToSvg() and serializeAsPngBlob() are helpers that I extracted out of the "export" dialog's machinery, but they are still in use there. I would like to give them a clean, unified interface, but it's definitely going to take a bit of thought.

Now that I've made this spike, and I understand the vague shape I'm looking for, I will look at how they're tested, and do a proper test-driven refactor as a separate PR.

**Comment by alsuren** [comment]
superseded by #3645

**Pull request closed without merging**